### PR TITLE
Implement engine_signalSuperchainV1

### PR DESCRIPTION
**Pull request by mininny**
Implementation for engine_signalSuperchainV1 method. 
RPC Spec: See https://github.com/ethereum-optimism/specs/blob/main/specs/protocol/exec-engine.md#engine_signalsuperchainv1

> Optional extension to the Engine API. Signals superchain information to the Engine: V1 signals which protocol version is recommended and required.
Introduces `rollup.halt` run flag. 
- `none`, default
- `patch`
- `minor`
- `major`

When `rollup.halt` is set to a specific version diff-level, the node will halt when the version difference is more *significant* than the diff-level specified. More specific examples of this behavior is explained in the testing section below.   

When `rollup.halt` is not set, but there is a significant version difference, the node will leave logs with corresponding severity as well.  

Based on https://github.com/ethereum-optimism/op-geth/pull/128 with minimal changes. 

---

## Testing

I haven't found any traces of Engine API testing in the op-erigon repo, so I did manual testings by triggering `engine_signalSuperchainV1` under different circumstances. 

Start Node: RPC running at port 8545:
```bash
./build/bin/erigon --datadir=datadir --chain=op-sepolia --http.port=8751 --authrpc.port=8545 --torrent.port=42068 --no-downloader --nodiscover --private.api.addr=127.0.0.1:9091 --http --ws --http.api=admin,engine,eth --p2p.allowed-ports=30309,30310,30311  --authrpc.jwtsecret=../secrets/jwt.hex  --p2p.protocol=67,68  --log.console.verbosity=3
```

### Scenario 1: Up-to-date engine protocol version
Node version: 3.1.1
Required version: 3.0.0
Recommended version: 3.1.0

Result is same for up-to-date engine for all flags

- none/patch/minor/major 
```bash
./build/bin/erigon OTHER_FLAGS --rollup.halt=none|patch|minor|major
```
RPC Sent: 
```
{
    "jsonrpc": "2.0",
    "method": "engine_signalSuperchainV1",
    "params": [
        {
            "required": "0x0000000000000000000000000000000000000003000000000000000000000000", // Represents 3.0.0 
            "recommended": "0x0000000000000000000000000000000000000003000000010000000000000000" // Represents 3.1.0
        }   
    ],
    "id": 1,
}
```

Result: 
```
// RPC Response
{'jsonrpc': '2.0', 'id': 1, 'result': '0x0000000000000000000000000000000000000003000000010000000000000001'}
// Node Log
None
```

### Scenario 2: Recommanded patch-outdated engine protocol version
Node version: 3.1.1
Required version: 3.1.0
Recommended version: 3.1.2

- patch/minor/major
```bash
./build/bin/erigon OTHER_FLAGS --rollup.halt=patch|minor|major
```
RPC Sent: 
```
{
    "jsonrpc": "2.0",
    "method": "engine_signalSuperchainV1",
    "params": [
        {
            "required": "0x0000000000000000000000000000000000000003000000010000000000000000", // Represents 3.1.0 
            "recommended": "0x0000000000000000000000000000000000000003000000010000000200000000" // Represents 3.1.2
        }   
    ],
    "id": 1,
}
```

Result: 
```
// RPC Response
{'jsonrpc': '2.0', 'id': 1, 'result': '0x0000000000000000000000000000000000000003000000010000000000000001'}
// Node Log
INFO[03-11|23:13:19.922] Outdated with support backward-compatible recommended protocol change local=v3.1.0-1 required=v3.1.0 recommended=v3.1.2
```

### Scenario 3: Required patch-outdated engine protocol version
Node version: 3.1.1
Required version: 3.1.2
Recommended version: 3.1.2

- patch
```bash
./build/bin/erigon OTHER_FLAGS --rollup.halt=patch
```
RPC Sent: 
```
{
    "jsonrpc": "2.0",
    "method": "engine_signalSuperchainV1",
    "params": [
        {
            "required": "0x0000000000000000000000000000000000000003000000010000000200000000", // Represents 3.1.2
            "recommended": "0x0000000000000000000000000000000000000003000000010000000200000000" // Represents 3.1.2
        }   
    ],
    "id": 1,
}
```

Result: 
```
// RPC Response
http.client.RemoteDisconnected: Remote end closed connection without response
// Node Log
INFO[03-11|23:16:43.416] Outdated with support backward-compatible required protocol change local=v3.1.0-1 required=v3.1.2 recommended=v3.1.2
EROR[03-11|23:16:43.416] Opted to halt, unprepared for protocol change required=v3.1.2 local=v3.1.0-1
INFO[03-11|23:16:43.416] Exiting...
INFO[03-11|23:16:43.416] Exiting Engine...
INFO[03-11|23:16:43.416] HTTP endpoint closed                     url=127.0.0.1:8751
INFO[03-11|23:16:43.416] RPC server shutting down
```

- minor/major
Halting is ignored for rollup.halt=minor/major when the version diff is only patch-level. 

```bash
./build/bin/erigon OTHER_FLAGS --rollup.halt=minor|major
```
RPC Sent: 
```
{
    "jsonrpc": "2.0",
    "method": "engine_signalSuperchainV1",
    "params": [
        {
            "required": "0x0000000000000000000000000000000000000003000000010000000200000000", // Represents 3.1.2
            "recommended": "0x0000000000000000000000000000000000000003000000010000000200000000" // Represents 3.1.2
        }   
    ],
    "id": 1,
}
```

Result: 
```
// RPC Response
{'jsonrpc': '2.0', 'id': 1, 'result': '0x0000000000000000000000000000000000000003000000010000000000000001'}
// Node Log
INFO[03-11|23:17:29.527] Outdated with support backward-compatible recommended protocol change local=v3.1.0-1 required=v3.1.2 recommended=v3.1.2
INFO[03-11|23:17:29.527] Outdated with support backward-compatible required protocol change local=v3.1.0-1 required=v3.1.2 recommended=v3.1.2
```

### Scenario 4: Required minor-outdated engine protocol version
Node version: 3.1.1
Required version: 3.2.0
Recommended version: 3.2.0

- patch/minor
```bash
./build/bin/erigon OTHER_FLAGS --rollup.halt=patch|minor
```
RPC Sent: 
```
{
    "jsonrpc": "2.0",
    "method": "engine_signalSuperchainV1",
    "params": [
        {
            "required": "0x0000000000000000000000000000000000000003000000020000000000000000", // Represents 3.2.0
            "recommended": "0x0000000000000000000000000000000000000003000000020000000000000000" // Represents 3.2.0
        }   
    ],
    "id": 1,
}
```

Result: 
```
// RPC Response
http.client.RemoteDisconnected: Remote end closed connection without response
// Node Log
WARN[03-11|23:20:39.938] Outdated with minor backward-compatible recommended protocol change local=v3.1.0-1 required=v3.2.0 recommended=v3.2.0
WARN[03-11|23:20:39.939] Outdated with minor backward-compatible required protocol change local=v3.1.0-1 required=v3.2.0 recommended=v3.2.0
EROR[03-11|23:20:39.939] Opted to halt, unprepared for protocol change required=v3.2.0 local=v3.1.0-1
INFO[03-11|23:20:39.939] Exiting...
INFO[03-11|23:20:39.939] HTTP endpoint closed                     url=127.0.0.1:8751
INFO[03-11|23:20:39.939] Exiting Engine...
INFO[03-11|23:20:39.939] RPC server shutting down
```

- major
Halting is ignored for rollup.halt=major when the version diff is only minor-level. 

```bash
./build/bin/erigon OTHER_FLAGS --rollup.halt=major
```
RPC Sent: 
```
{
    "jsonrpc": "2.0",
    "method": "engine_signalSuperchainV1",
    "params": [
        {
            "required": "0x0000000000000000000000000000000000000003000000020000000000000000", // Represents 3.2.0
            "recommended": "0x0000000000000000000000000000000000000003000000020000000000000000" // Represents 3.2.0
        }   
    ],
    "id": 1,
}
```

Result: 
```
// RPC Response
{'jsonrpc': '2.0', 'id': 1, 'result': '0x0000000000000000000000000000000000000003000000010000000000000001'}
// Node Log
WARN[03-11|23:21:26.808] Outdated with minor backward-compatible recommended protocol change local=v3.1.0-1 required=v3.2.0 recommended=v3.2.0
WARN[03-11|23:21:26.808] Outdated with minor backward-compatible required protocol change local=v3.1.0-1 required=v3.2.0 recommended=v3.2.0
```

### Scenario 5: Required major-outdated engine protocol version
Node version: 3.1.1
Required version: 4.0.0
Recommended version: 4.0.0

- patch/minor/major
```bash
./build/bin/erigon OTHER_FLAGS --rollup.halt=patch|minor|major
```
RPC Sent: 
```
{
    "jsonrpc": "2.0",
    "method": "engine_signalSuperchainV1",
    "params": [
        {
            "required": "0x0000000000000000000000000000000000000004000000000000000000000000", // Represents 4.0.0
            "recommended": "0x0000000000000000000000000000000000000004000000000000000000000000" // Represents 4.0.0
        }   
    ],
    "id": 1,
}
```

Result: 
```
// RPC Response
http.client.RemoteDisconnected: Remote end closed connection without response
// Node Log
EROR[03-11|23:23:35.265] Outdated with major recommended protocol change local=v3.1.0-1 required=v4.0.0 recommended=v4.0.0
EROR[03-11|23:23:35.265] Outdated with major required protocol change local=v3.1.0-1 required=v4.0.0 recommended=v4.0.0
EROR[03-11|23:23:35.265] Opted to halt, unprepared for protocol change required=v4.0.0 local=v3.1.0-1
INFO[03-11|23:23:35.265] Exiting...
INFO[03-11|23:23:35.265] Exiting Engine...
INFO[03-11|23:23:35.265] RPC server shutting down
INFO[03-11|23:23:35.265] HTTP endpoint closed                     url=127.0.0.1:8751
```

- none
Halting is ignored for rollup.halt=none.  

```bash
./build/bin/erigon OTHER_FLAGS --rollup.halt=none
```
RPC Sent: 
```
{
    "jsonrpc": "2.0",
    "method": "engine_signalSuperchainV1",
    "params": [
        {
            "required": "0x0000000000000000000000000000000000000004000000000000000000000000", // Represents 4.0.0
            "recommended": "0x0000000000000000000000000000000000000004000000000000000000000000" // Represents 4.0.0
        }   
    ],
    "id": 1,
}
```

Result: 
```
// RPC Response
{'jsonrpc': '2.0', 'id': 1, 'result': '0x0000000000000000000000000000000000000003000000010000000000000001'}
// Node Log
EROR[03-11|23:24:08.871] Outdated with major recommended protocol change local=v3.1.0-1 required=v4.0.0 recommended=v4.0.0
EROR[03-11|23:24:08.871] Outdated with major required protocol change local=v3.1.0-1 required=v4.0.0 recommended=v4.0.0
```

### CODE REVIEW

File: turbo/engineapi/engine_server.go [hunk 1]
```diff
@@ -5,6 +5,8 @@ import (
 	"encoding/binary"
 	"errors"
 	"fmt"
+	"github.com/ledgerwatch/erigon/eth/ethconfig"
+	"github.com/ledgerwatch/erigon/params"
```
Comment (ImTei) [inline turbo/engineapi/engine_server.go]: This will cause lint CI to fail. you should reorder import lines.

File: cmd/utils/flags.go [hunk 2]
```diff
@@ -1887,6 +1891,10 @@ func SetEthConfig(ctx *cli.Context, nodeConfig *nodecfg.Config, cfg *ethconfig.C
 			cfg.DisableTxPoolGossip = ctx.Bool(RollupDisableTxPoolGossipFlag.Name)
 		}
 	}
+
+	if ctx.IsSet(RollupHaltOnIncompatibleProtocolVersionFlag.Name) {
+		cfg.RollupHaltOnIncompatibleProtocolVersion = ctx.String(RollupHaltOnIncompatibleProtocolVersionFlag.Name)
```
Comment (ImTei) [inline cmd/utils/flags.go]: why don't you do sanity check for the flag value?

**Review by ImTei — COMMENTED** [review]
Not yet reviewed detailed functionalities, but left some comments.

File: turbo/engineapi/engine_server.go (hunk 1 above)
Comment (pcw109550) [reply turbo/engineapi/engine_server.go]: `gofmt -w -s [FILENAME]`

File: cmd/utils/flags.go (hunk 2 above)
Comment (mininny) [reply cmd/utils/flags.go]: I added the sanity check here: 1fda97c20c79d2c883693c0884204b119b00a75a

File: turbo/engineapi/engine_server.go (hunk 1 above)
Comment (mininny) [reply turbo/engineapi/engine_server.go]: fecdc47b357a37d80349c9a6e23bcd6d5a1d6f62

File: params/superchain.go [hunk 3]
```diff
@@ -37,6 +39,8 @@ var (
 	chaosnetRegolithTime = big.NewInt(1692156862)
 )
 
+var OPStackSupport = ProtocolVersionV0{Build: [8]byte{}, Major: 3, Minor: 1, Patch: 0, PreRelease: 1}.Encode()
```
Comment (ImTei) [inline params/superchain.go]:
```suggestion
var OPStackSupport = ProtocolVersionV0{Build: [8]byte{}, Major: 6, Minor: 0, Patch: 0, PreRelease: 0}.Encode()
```
https://github.com/ethereum-optimism/op-geth/pull/243/files#diff-7663c343438d424a4d4046a455bf01862dea827d55849b3ac95a388c57c3899cR14

Current op-erigon supports Ecotone. need to update protocol version.

**Review by ImTei — CHANGES_REQUESTED** [review]
I also manually tested RPC and it works!

File: params/superchain.go (hunk 3 above)
Comment (mininny) [reply params/superchain.go]: Thanks, I updated it to 6.0.0 :+1:

**Review by ImTei — APPROVED** [review]

**Pull request merged**